core/primitives/src/stateless_validation/chunk_endorsements_bitmap.rs

```diff
@@ -0,0 +1,119 @@
+use bitvec::prelude::*;
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_primitives_core::types::ShardId;
+use near_schema_checker_lib::ProtocolSchema;
+
+/// Represents a collection of bitmaps, one per shard, to store whether the endorsements from the chunk validators has been received.
+///
+/// For each shard, the endorsements are encoded as a sequence of bits: 1 means endorsement received and 0 means not received.
+/// While the number of chunk validator seats is fixed, the number of chunk-validator assignments may be smaller and may change,
+/// since the seats are assigned to validators weighted by their stake. Thus, we represent the bits as a vector of bytes.
+/// The number of assignments may be less or equal to the number of total bytes. This representation allows increasing
+/// the chunk validator seats in the future (which will be represented by a vector of greater length).
+#[derive(
+    BorshSerialize,
+    BorshDeserialize,
+    serde::Serialize,
+    serde::Deserialize,
+    Debug,
+    Clone,
+    Eq,
+    PartialEq,
+    Default,
+    ProtocolSchema,
+)]
+pub struct ChunkEndorsementsBitmap {
+    inner: Vec<Vec<u8>>,
+}
+
+/// Type of BitVec used in the internal implementation.
+type BitVecType = BitVec<u8, Lsb0>;
+
+impl ChunkEndorsementsBitmap {
+    /// Creates an empty endorsement bitmap for each shard.
+    pub fn new(num_shards: usize) -> Self {
+        Self { inner: vec![Default::default(); num_shards] }
+    }
+
+    // Creates an endorsement bitmap for all the shards.
+    pub fn from_endorsements(shards_to_endorsements: Vec<Vec<bool>>) -> Self {
+        let mut bitmap = ChunkEndorsementsBitmap::new(shards_to_endorsements.len());
+        for (shard_id, endorsements) in shards_to_endorsements.into_iter().enumerate() {
+            bitmap.add_endorsements(shard_id as ShardId, endorsements);
+        }
+        bitmap
+    }
+
+    /// Adds the provided endorsements to the bitmap for the specified shard.
+    pub fn add_endorsements(&mut self, shard_id: ShardId, endorsements: Vec<bool>) {
+        let bitvec: BitVecType = endorsements.iter().collect();
+        self.inner[shard_id as usize] = bitvec.into();
+    }
+
+    /// Returns an iterator over the endorsements (yields true if the endorsement for the respective position was received).
+    pub fn iter(&self, shard_id: ShardId) -> Box<dyn Iterator<Item = bool>> {
+        let bitvec = BitVecType::from_vec(self.inner[shard_id as usize].clone());
+        Box::new(bitvec.into_iter())
+    }
```

Comment:
Note for the future self: let's not forget to validate bitmap properly, including that length can't be ridiculously high and that latest bits must be set to 0.

Reply:
Added this as a task to the issue.